### ISSUE TITLE
p::d::SolutionTransfer::deserialize() fails on non-coarse meshes

### ISSUE THREAD

**Issue by elauksap** [dealii/dealii#11768]
Hi everybody,

the [attached code](https://github.com/dealii/dealii/files/5997690/serialization.zip) (inspired by [this test](https://github.com/dealii/dealii/blob/master/tests/fullydistributed_grids/solution_transfer_01.cc)) fails if run in parallel.

It shows a simple use of `p::d::SolutionTransfer::{serialize,deserialize}()` for de-serializing a non-coarse `p::d::Triangulation`.

Success or failure depends on the number of MPI cores (on my machine it fails with 6 cores).
The problem arises both in 2D and in 3D, using either deal.II or Trilinos or PETSc distributed vectors and on older deal.II releases.

The code works correctly for `p::f::Triangulation`s or if the serialized triangulation is coarse.

Any clues?

Thanks!

Attachment:
[serialization.zip](https://github.com/dealii/dealii/files/5997690/serialization.zip)

**Comment by marcfehling** [comment]
What exactly is your error message?

We have tests that demonstrate how to use it with `p:d::Tria` with `p::d::SolutionTransfer::{serialize,deserialize}()`. I'm not familiar with the differences between `distributed` and `fullydistributed` applications in this regard. See [p4est_save_02.cc](https://github.com/dealii/dealii/blob/master/tests/mpi/p4est_save_02.cc), [p4est_save_03.cc](https://github.com/dealii/dealii/blob/master/tests/mpi/p4est_save_03.cc), [p4est_save_04.cc](https://github.com/dealii/dealii/blob/master/tests/mpi/p4est_save_04.cc), [p4est_save_06.cc](https://github.com/dealii/dealii/blob/master/tests/mpi/p4est_save_06.cc).

**Comment by marcfehling** [comment]
You are using a ghosted vector for deserialization in your example. You should use one without ghost elements (see tests and [doc](https://www.dealii.org/developer/doxygen/deal.II/classparallel_1_1distributed_1_1SolutionTransfer.html#a4237d8449184deebd4fcb99891ae9c69)).

Change line
```
solution_transfer.deserialize(vector);
```
to
```
solution_transfer.deserialize(vector_owned);
```
and I bet it will work. The following line `vector = vector_owned;` will make sense then, too ;-)

**Comment by elauksap** [comment]
Thanks for having a look @marcfehling!

Sorry for the `solution_transfer.deserialize(vector_owned);`, it was just a typo but unfortunately this does not solve the problem.

This is the error I get:

> An error occurred in line <1644> of file </home/elauksap/software/dealii/include/deal.II/lac/la_parallel_vector.h> in function
>     Number& dealii::LinearAlgebra::distributed::Vector< <template-parameter-1-1>, <template-parameter-1-2> >::operator()(dealii::LinearAlgebra::distributed::Vector< <template-parameter-1-1>, <template-parameter-1-2> >::size_type) [with Number = double; MemorySpace = dealii::MemorySpace::Host; dealii::LinearAlgebra::distributed::Vector< <template-parameter-1-1>, <template-parameter-1-2> >::size_type = unsigned int]
> The violated condition was:
>     partitioner->in_local_range(global_index) || partitioner->ghost_indices().is_element(global_index)
> Additional information:
>     You tried to access element 4294967295 of a distributed vector, but
>     this element is not stored on the current processor. Note: The range
>     of locally owned elements is [4193,4912], and there are 1017 ghost
>     elements that this vector can access.

I've compared my code with the other tests in deal.II but couldn't find any possible source for the above error.

Thanks a lot again! Any further hint is appreciated 😄

**Comment by marcfehling** [comment]
I'm sure `4294967295` corresponds to `numbers::invalid_unsigned_int`. You didn't call `distribute_dofs()` after loading the triangulation. Correspondingly, you did not update your IndexSets, nor did you initialize your vectors with those. Could that be the issue?

At this point, it's like finding a needle in a haystack. Have you tried using a debugger to locate the error any further?

**Comment by marcfehling** [comment]
The `DoFHandler` will be re-initialized when the corresponding `Triangulation` will be created since #11627. This means also the enumeration of DoFs will be gone.

I noticed we should probably clear the `DoFHandler` also when `Triangulation::clear()` is called. I will provide a PR.

This will not solve your issue though :) You need to follow the steps in the comment before to make it work.

**Comment by marcfehling** [comment]
I think the reason why your reference test works and your test doesn't, is because you reinitialize the coarse triangulation with `GridGenerator::hyper_cube(triangulation);`, but the reference test doesn't.

We reinit associated `DoFHandlers` with the creation of a `Triangulation`, but don't clear them when clearing the `Triangulation`, which we should in my opinion. I'm sure my proposed PR will break your reference test.

**Comment by elauksap** [comment]
@marcfehling thanks a lot! It seems that you got the point.

The difference between the test attached here and the reference test is the use of `p::d::T` against `p::f::T`. Indeed `p::d::T::load()` assumes that a coarse mesh exists, wherease `p::f::T::load()` assumes that the triangulation is empty.
This is why here I re-initialize the triangulation with `GridGenerator::hyper_cube()`.

It doesn't seem to be related to #11798 because the test failed in release 9.2.0, too.

Anyway, adding `dof_handler.distribute_dofs(...);` after `triangulation.load()` as you suggested seems to fix the issue!

It is not clear to me why we need to re-distribute the dofs again, even if the initial and the final meshes are the same and have the same parallel partitioning. Could you please explain the rationale behind this concept? Do I need to re-initialize the `IndexSet`s as well?

Thanks a lot again, your comments were really helpful!

**Comment by marcfehling** [comment]
> It is not clear to me why we need to re-distribute the dofs again, even if the initial and the final meshes are the same and have the same parallel partitioning. Could you please explain the rationale behind this concept? Do I need to re-initialize the `IndexSet`s as well?

I have no idea right now. I try to get a picture of it myself.

**Comment by marcfehling** [comment]
The time you save and load data to the file system, you typically interrupt your program and exit it entirely so that you can resume it later. At this time, also all information about enumerated degrees of freedom and `IndexSets` will be lost, so you will need to recreate them after calling `load`.

The tests in the testsuite do not get interrupted in the intended way of serialization as we need to verify the results in the same breath we create them. Calling `clear` and then `load` on the same `Triangulation` instance does not resemble the real use case. In fact, you uncovered a bug: whenever clearing a `Triangulation`, the information on associated `DoFHandlers` are invalid and should be cleared as well. The proposed PR, that will close this issue, ensures exactly that.

Long story short: You should `distribute_dofs` after deserialization. It might work for that one specific case in your reference test, but will most definitely not in other use cases and will leave others with headaches (including me :-D).

**Comment by elauksap** [comment]
The issue is solved. Thank you very much @marcfehling! You were really helpful. 🙏
